### PR TITLE
Adding an option to save pokemon spawns

### PR DESCRIPTION
**Pull request by dmateusp**
## Feature description:
Everytime a pokemon is encountered, dumps in `data/pokemon_spawns-USERNAME.json` the location and the pokemon name encountered.

example : `{"pokemon": "Krabby", "location": "53.3373282799,-6.26015253492"}`

## Why is this interesting ?
Having the spawns stored this way have these advantages:

- Do not require additional calls
- Is specific to one bot and its location
- Can be used by other modules and analytics to pin these spawns to web maps and plan efficient routes

EDIT : time has been added as per @MZorzy comment, pokemons spawn for a certain duration so the time when the pokemon was encountered is valuable information
`{"pokemon": "Poliwag", "location": "53.3369929597,-6.27457081671", "time": "18:48:19"}`

### CODE REVIEW

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 1]
```diff
@@ -453,3 +457,34 @@ def generate_throw_quality_parameters(self, throw_parameters):
...
+
+                self.emit_event(
+                    'save_spawn',
+                    formatted='Spawn at {location} saved in '+ file_name +' (Encountered {pokemon})',
+                    data={
+                        'pokemon': pokemon.name,
+                        'location': str(self.pokemon['latitude'])+','+str(self.pokemon['longitude'])
+                    }
+                )
+            except:
```
Comment (cwild) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: You should at the very least only capture ```Exception```.

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 2]
```diff
@@ -453,3 +457,34 @@ def generate_throw_quality_parameters(self, throw_parameters):
         throw_parameters['normalized_reticle_size'] = 1.25 + 0.70 * random()
         throw_parameters['normalized_hit_position'] = 0.0
         throw_parameters['throw_type_label'] = 'Normal'
+
+    def dump_pokemon_spawn(self, pokemon):
+        save_pokemon_spawn = self.config.save_pokemon_spawn
+        if save_pokemon_spawn:
+            file_name = 'pokemon_spawns-%s.json' % self.config.username
```
Comment (BriceSD) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: You should add this file to the .gitignore

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py (hunk 1 above)
Comment (dmateusp) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: sorry could you be more specific I'm not sure I understand ?

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py (hunk 2 above)
Comment (dmateusp) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: @BriceSD seems like I forgot, just fixed thanks

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py (hunk 1 above)
Comment (dmateusp) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: is my last change ok for you ? @cwild

**Comment by MZorzy** [comment]
Look here
https://github.com/PokemonGoMap/PokemonGo-Map/pull/390

Spawn point id is the coordinate.

**Comment by dmateusp** [comment]
@MZorzy so it's actually already done?

**Comment by MZorzy** [comment]
@dmateusp not here, but i think is a better idea to save ID (can be extracted lat and long) and add TIME in second (after hour) to spawn, like other start to do on MAP/other bot.
Look (not) all spawn are 15 min long, then 45min hidden, so  better keep this time info for efficent plan routes

**Comment by dmateusp** [comment]
@MZorzy got you! I will implement that :)

**Comment by dmateusp** [comment]
`{"pokemon": "Poliwag", "location": "53.3369929597,-6.27457081671", "time": "18:48:19"}` is the new format, I'm not sure what @MZorzy meant by spawn ID

**Comment by TheSavior** [comment]
My understanding is that @douglascamata wants to support this with the websockets and wants to remove the file writing stuff from the app. Hopefully he can review this and talk about how best to proceed

**Comment by dmateusp** [comment]
@TheSavior I understand, maybe I could redo this with web sockets but I'll need to have a better understanding of it before ^^

**Comment by douglascamata** [comment]
@dmateusp why keep writing files for information when we want to get rid of all of them?

**Comment by TheSavior** [comment]
@douglascamata Perhaps we need a better example / documentation of how another tool can hook into the event system to get this kind of information.

**Comment by dmateusp** [comment]
@douglascamata @TheSavior I did not know we were moving away from dumping json files, I thought the web socket was still a work in progress, do you want me to close my pull request and work on web sockets instead ?

**Comment by TheSavior** [comment]
If you can, yes. Everything should be available to you already since we send web socket events for pretty much everything.

**Comment by douglascamata** [comment]
@dmateusp contact me in Slack so we can discuss about creating some good docs for the event system and how ppl can listen to events on the websocket server.

**Comment by dmateusp** [comment]
@douglascamata I will once I get some time :) !

**Comment by k4n30** [comment]
@dmateusp Closing this PR?

**Comment by solderzzc** [comment]
Need solve the conflict.

**Comment by BreezeRo** [comment]
whats the status?

**Comment by solderzzc** [comment]
close since no  activity in 7 days/

**Comment by MZorzy** [comment]
this pull look good,  we can create a map path with (more or less) sure spawn time. and miss no one.  
please fix it.

**Comment by dmateusp** [comment]
Sorry everyone, been busy days at work. I solved the issues but for some reason I can't reopen the PR ?

**Comment by Anakin5** [comment]
looks like this stuff has been merge and it breaks everything...

**Comment by dmateusp** [comment]
has it ?

**Comment by MZorzy** [comment]
17.711 addittion, 1.413 deletion.,really ??

**Comment by Anakin5** [comment]
needs to be reverted

**Comment by dmateusp** [comment]
I see +44 −1 , I don't see it merged ?

**Comment by Anakin5** [comment]
It is there , on head of dev branch 838fea42

**Pull request closed without merging**